File: Validate Binary Search Tree/Validate Binary Search Tree.rb
# Definition for a binary tree node.
class TreeNode
    attr_accessor :val, :left, :right
    def initialize(val = 0, left = nil, right = nil)
        @val = val
        @left = left
        @right = right
    end
end

def is_valid_bst(root)
    valid_bst_helper(root, -Float::INFINITY, Float::INFINITY)
end

def valid_bst_helper(node, lower, upper)
    return true if node.nil?

    if node.val <= lower || node.val >= upper
        return false
    end

    valid_bst_helper(node.left, lower, node.val) && valid_bst_helper(node.right, node.val, upper)
end

# Example usage
root = TreeNode.new(5)
root.left = TreeNode.new(1)
root.right = TreeNode.new(4)
root.right.left = TreeNode.new(3)
root.right.right = TreeNode.new(6)

puts is_valid_bst(root) # Output: false


# Example usage
root = TreeNode.new(2)
root.left = TreeNode.new(1)
root.right = TreeNode.new(3)

puts is_valid_bst(root) # Output: true

File: Validate Binary Search Tree/Validate Binary Search Tree.js
function isValidBST(root) {
    return isValidBSTHelper(root, -Infinity, Infinity);
}

function isValidBSTHelper(node, lower, upper) {
    if (node === null) {
        return true;
    }

    if (node.val <= lower || node.val >= upper) {
        return false;
    }

    return isValidBSTHelper(node.left, lower, node.val) && isValidBSTHelper(node.right, node.val, upper);
}

// Example usage
const root = {
    val: 5,
    left: {
        val: 1,
        left: null,
        right: null
    },
    right: {
        val: 4,
        left: {
            val: 3,
            left: null,
            right: null
        },
        right: {
            val: 6,
            left: null,
            right: null
        }
    }
};

console.log(isValidBST(root)); // Output: false

// Example usage2
const root2 = {
    val: 2,
    left: {
        val: 1,
        left: null,
        right: null
    },
    right: {
        val: 3,
        left: null,
        right: null
    }
};

console.log(isValidBST(root2)); // Output: true
